src/core/file_descriptor.c

```diff
@@ -0,0 +1,455 @@
+//
+// Copyright 2020 Staysail Systems, Inc. <info@staysail.tech>
+// Copyright 2018 Capitar IT Group BV <info@capitar.com>
+// Copyright 2019 Devolutions <info@devolutions.net>
+//
+// This software is supplied under the terms of the MIT License, a
+// copy of which should be located in the distribution where this
+// file was obtained (LICENSE.txt).  A copy of the license may also be
+// found online at https://opensource.org/licenses/MIT.
+//
+
+#include <stdint.h>
+#include <string.h>
+
+#include <nng/nng.h>
+
+#include "core/nng_impl.h"
+#include "core/file_descriptor.h"
+
+typedef struct {
+	nng_stream_dialer ops;
+	int               fd;
+  //	char *            host;
+  //	char *            port;
+	int               af; // address family
+	bool              closed;
+	nng_sockaddr      sa;
+	nni_file_descriptor_dialer *  d;      // platform dialer implementation
+	nni_aio *         resaio; // resolver aio
+	nni_aio *         conaio; // platform connection aio
+	nni_list          conaios;
+	nni_mtx           mtx;
+} file_descriptor_dialer;
+
+static void
+file_descriptor_dial_cancel(nni_aio *aio, void *arg, int rv)
+{
+	file_descriptor_dialer *d = arg;
+
+	nni_mtx_lock(&d->mtx);
+	if (nni_aio_list_active(aio)) {
+		nni_aio_list_remove(aio);
+		nni_aio_finish_error(aio, rv);
+
+		if (nni_list_empty(&d->conaios)) {
+			nni_aio_abort(d->conaio, NNG_ECANCELED);
+			nni_aio_abort(d->resaio, NNG_ECANCELED);
+		}
+	}
+	nni_mtx_unlock(&d->mtx);
+}
+
```

Comment:
The name resolution parts of TCP should be thrown away-- I forgot about that when we discussed.